code/ui/components/src/new/Button/Button.tsx

```diff
@@ -1,44 +1,34 @@
-import type { ReactNode } from 'react';
 import React, { forwardRef } from 'react';
 import { styled } from '@storybook/theming';
 import { darken, lighten, rgba, transparentize } from 'polished';
+import type { Icons } from '@storybook/icons';
 import type { PropsOf } from '../utils/types';
+import { Icon } from '../Icon/Icon';
 
-interface CommonProps<T extends React.ElementType = React.ElementType> {
+interface ButtonProps<T extends React.ElementType = React.ElementType> {
+  children: string;
   as?: T;
   size?: 'small' | 'medium';
-  variant?: 'primary' | 'secondary' | 'tertiary';
+  variant?: 'solid' | 'outline' | 'ghost';
   onClick?: () => void;
   disabled?: boolean;
   active?: boolean;
+  icon?: Icons;
 }
 
-type ButtonIconOnlyProps = {
-  iconOnly: true;
-  icon: ReactNode;
-  children?: never;
-};
-
-type ButtonWithTextProps = {
-  iconOnly?: false;
-  icon?: ReactNode;
-  children: string;
-};
-
-type ButtonProps<T extends React.ElementType = React.ElementType> = CommonProps<T> &
-  (ButtonIconOnlyProps | ButtonWithTextProps);
-
 export const Button: {
   <E extends React.ElementType = 'button'>(
     props: ButtonProps<E> & Omit<PropsOf<E>, keyof ButtonProps>
   ): JSX.Element;
   displayName?: string;
 } = forwardRef(
   ({ as, children, icon, ...props }: ButtonProps, ref: React.Ref<HTMLButtonElement>) => {
+    const LocalIcon = Icon[icon];
+
     return (
       <StyledButton as={as} ref={ref} {...props}>
-        {icon}
-        {!props.iconOnly && children}
+        {icon && <LocalIcon />}
```

Comment:
Actually, can we not just tell users to put the icon themself before the children?

Reply:
Not sure I follow. We want to control the look of the icon as much as possible. If we allow users to put the icon in `children` they can put anything, at any size, before or after the text.